contrib/storage-ipfs/src/main/java/org/apache/drill/exec/store/ipfs/IPFSCompat.java

```diff
@@ -0,0 +1,318 @@
+            }
+          },
+          executor
+      );
+      if (ret.get().size() > 0) {
+        return ret.get();
+      } else {
+        return Collections.emptyList();
+      }
+    }
+  }
+
+  public class Name {
+    /**
+     * Resolve a IPNS name.
+     * See <a href="https://docs.ipfs.io/reference/http/api/#api-v0-name-resolve">name/resolve in IPFS doc</a>.
+     * @param hash the IPNS name to resolve
+     * @param timeout timeout value in seconds
+     * @param executor executor
+     * @return a Multihash of resolved name
+     */
+    public Optional<String> resolve(Multihash hash, int timeout, ExecutorService executor) {
+      AtomicReference<String> ret = new AtomicReference<>();
+      timeLimitedExec(
+        "name/resolve?arg=" + hash,
+        timeout,
+        res -> {
+          Map peer = (Map) res;
+          if (peer != null) {
+            ret.set((String) peer.get(("Path")));
+            return true;
+          }
+          return false;
+        },
+        err -> {
+          if (!(err instanceof TimeoutException)) {
+            throw new RuntimeException(err);
+          }
+        },
+        executor
+      );
+      return Optional.ofNullable(ret.get());
+    }
+  }
+
+  private void timeLimitedExec(String path, int timeout, Predicate<Object> processor, Consumer<Exception> error,
+                               ExecutorService executor) {
+    CompletableFuture<Void> f = CompletableFuture.runAsync(
+      ()-> getObjectStream(path, processor, error),
+      executor
+    );
+    try {
+      f.get(timeout, TimeUnit.SECONDS);
+    } catch (TimeoutException | ExecutionException | InterruptedException e) {
+      f.cancel(true);
+      error.accept(e);
+    }
+  }
+
+  private void getObjectStream(String path, Predicate<Object> processor, Consumer<Exception> error) {
+    byte LINE_FEED = (byte)10;
+
+    try {
+      InputStream in = getStream(path);
+      ByteArrayOutputStream resp = new ByteArrayOutputStream();
+
+      byte[] buf = new byte[4096];
+      int r;
+      while ((r = in.read(buf)) >= 0) {
+        resp.write(buf, 0, r);
+        if (buf[r - 1] == LINE_FEED) {
+          try {
+            boolean done = processor.test(JSONParser.parse(resp.toString()));
+            if (done) {
+              break;
+            }
+            resp.reset();
+          } catch (IllegalStateException e) {
+            in.close();
+            resp.close();
+            error.accept(e);
+          }
+        }
+      }
+      in.close();
+      resp.close();
+    } catch (IOException e) {
+      error.accept(e);
+    }
+  }
+
+  private InputStream getStream(String path) throws IOException {
+    URL target = new URL(protocol, host, port, version + path);
+    HttpURLConnection conn = configureConnection(target, "POST", readTimeout);
+    return conn.getInputStream();
+  }
+
+  private static HttpURLConnection configureConnection(URL target, String method, int timeout) throws IOException {
```

Comment:
I'm getting an IDE warning here about `method` always being `POST`.  Can we make this a local variable?  Will it ever be `GET` or something else?

Reply:
Fixed in d2ea637.